### ISSUE TITLE
Payload format for meta operations

### ISSUE THREAD

**Issue by egekorkan** [w3c/wot-thing-description#1231]
I had mentioned this in #1200 but it applies to all meta operations. The payload formats of operations like `readallproperties` have responses from the Thing that contain the values of multiple affordances. There are multiple issues around this:
1. How the payload looks like is badly documented for these operations. There is only one "theoretical" sounding paragraph just **above** https://w3c.github.io/wot-thing-description/#interactionaffordance. There are no examples and there is no mention of operations like `subscribeallevents` or `unsubscribeallevents`.
2. Events and observable properties: Should the name of the affordance appear in the map of the payload? For example, if I am observing all properties but only 1 has changed (which is the likely case), should I still get a map with all the property names (unchanged ones will be empty?) or should their names be omitted from the map? I would have done the second one by intuition but the spec requires the first one. For example, the first one would look like this:
```json
{
  "myProp1": 12,
  "myProp2": null
}
```
3. What happens when some of the affordances have non JSON payloads?
4. (small point) The following text is not clear:
> ObjectSchema instance contains each data schema of the PropertyAffordances

I think that it should not say `each data schema` but `payload`?

**Comment by benfrancis** [comment]
> 1. How the payload looks like is badly documented for these operations. There is only one "theoretical" sounding paragraph just **above** https://w3c.github.io/wot-thing-description/#interactionaffordance.

I agree this part of the spec needs work. These normative statements are quite buried in the text and lack examples. Even though I know they are there I always have to dig for a while to find them.

> There are no examples and there is no mention of operations like `subscribeallevents` or `unsubscribeallevents`.

In the case of the Core Profile the `subscribeallevents` and `unsubscribeallevents` operations using Server-Sent Events don't actually have a payload so it's a bit of a non-issue https://w3c.github.io/wot-profile/#subscribeallevents

For implementations which do have a payload, can you think of a reasonable default payload schema constructed from the schemas of  `subscribeevent` and `unsubscribeevent` operations? Bearing in mind this schema is probably just for the subscription operation, not the actual event payloads (see below).

If something other than the default should be used, it's not clear to me where a schema should be defined. It's possible to define an `ExpectedResponse` in a top level `Form` and we have previously discussed adding a `schema` member to `ExpectedResponse` like the one in `AdditionalExpectedResponse`. There's currently no way to define a _request_ schema in a top level form though.

> 2. Events and observable properties: Should the name of the affordance appear in the map of the payload? For example, if I am observing all properties but only 1 has changed (which is the likely case), should I still get a map with all the property names (unchanged ones will be empty?) or should their names be omitted from the map? I would have done the second one by intuition but the spec requires the first one.

I think the nature of events and observed properties makes them different from reading and writing all properties in that when you subscribe to all events or observe all properties, you probably expect to receive individual events or property changes in each payload rather than all in one payload, since they probably won't occur at the same time. I would therefore assume that the payload of an event subscribed to via a `subscribeallevents` operation is the same as the payload of an event subscribed to via a `subscribeevent` operation. The one caveat is that in the case of a `subscribeallevents` operation it may be necessary to include the event name in the payload in order to distinguish between different event types. The same applies to observed properties.

For example, in the Core Profile an event payload sent via Server-Sent Events looks like this:

```
event: overheated\n
data: 90\n
id: 12345\n\n
```

In this case the event name is included in the event payload, regardless of whether it was subscribed to via a `subscribeevent` or `subscribeallevents` operation.

> 3. What happens when some of the affordances have non JSON payloads?

Presumably you'd specify a different `contentType` in the `Form`?

> 5. (small point) Why are these called meta-interactions? I think that meta-operations sound more correct?

I agree. Now that this section of the specification is growing beyond a single special case, perhaps there should be a new section called "Meta Operations"?

**Comment by benfrancis** [comment]
As discussed in #1200, the data schema for a `queryallactions` operation is also tricky. Like events and observed property changes they may be multiple instances of an action, but unlike events and observed property changes actions do not just occur at an instant in time but may be long-running, potentially with multiple actions queued up or running at the same time. Therefore you may expect to receive the status of multiple action instances in a single payload when querying "all actions".

Another difference from properties is that the payload of a `queryaction` operation may not necessarily match the `output` schema of the action affordance. When querying an ongoing action (which doesn't yet have an "output"), the response payload may have a different schema which represents the action status, like the [`ActionStatus`](https://w3c.github.io/wot-profile/#ActionStatus) object in the Core Profile. It might therefore make more sense for the default schema of a `queryallactions` operation to be constructed from the `response` members of the `Form`s of `queryaction` operations (if present), rather than the output data schema of the `ActionAffordance`.

That could either be an array of `queryaction` `response`s, or a map of arrays keyed by action name.

E.g. (from https://github.com/w3c/wot-thing-description/issues/1200#issuecomment-919211937):

```json
{
  "fade": [
    {
      "status": "completed",
      "href": "/things/lamp/actions/fade/123e4567-e89b-12d3-a456-426655"
    },
    {
      "status": "pending",
      "href": "/things/lamp/actions/fade/123e4567-e89b-12d3-a456-431553"
    },
    {
      "status": "pending",
      "href": "/things/lamp/actions/fade/123e4567-e89b-12d3-a456-51ff16"
    }
  ],
  "reboot": [
    {
      "status": "pending",
      "href": "/things/lamp/actions/reboot/123e4567-e89b-12d3-a456-f3dea"
    }
  ]
}
```

**Comment by egekorkan** [comment]
@benfrancis I will try to reply to your comments above:

> a new section called "Meta Operations"?

I would support this.

---

Regarding the other points, I think there are two things I need to clarify:
1. I was thinking of subscribing to an endpoint once that sends all events. Doing a longpoll to a /mything/events for example. In this case, I would not know which event I receive unless they are always encapsulated in an object with the event name. However, there are other cases that I did not think of. In vanilla MQTT, subscribing to mything/events/* would actually do subscriptions do all the possible events individually (by the broker). So once an event is delivered, it would be delivered to mything/events/myevent and I can distinguish without an object wrapper.
2. I was not talking about the subscription and subscription payloads. There is no way to understand what I had meant just by the operation name since receiving an event is not an operation. I should have been more specific.

> What happens when some of the affordances have non JSON payloads?
Presumably you'd specify a different contentType in the Form?

Below are some payloads of events that result from a `subscribeallevents` operation that can be received if I have all events delivering JSON payloads:
```json
{
  "myEvent1": 12,
  "myEvent2": 24
}
```
```json
{
  "myEvent1": 12
}
```
However, if one of the events send a JPEG that is not JSON encoded, what should happen?
```js
{
  "myEvent3": //somehow a JPEG
}
```

Of course, if all events are received like in the case of MQTT (or SSE in core profile but I am not 100% sure), this issue would not matter too much but this is not the case.

**Comment by benfrancis** [comment]
> However, if one of the events send a JPEG that is not JSON encoded, what should happen?

Surely by definition an event happens at an instance in time and therefore each event would usually be delivered separately? Under what circumstances would events be grouped together? I can imagine this being done as an optimisation for very high frequency events, but for the purposes of defining a default is it not a reasonable assumption that each event is delivered in a separate payload?

Depending on the delivery mechanism that still might not solve the problem of binary payloads though:
1. For Server-Sent Events all messages are UTF-8 encoded which means that if you want to send a binary file as part of a payload then you have no choice but to base64 encode it.
2. It is possible to send binary data over WebSockets (e.g. using an `ArrayBuffer` or `Blob` in JavaScript) but if you want to wrap that binary data in a JSON wrapper (e.g. to specify the event name and/or a timestamp) then you're probably still going to need to base64 encode the binary data
3. Any other mechanism like long-polling is going to have the same issue if you want to wrap the payload in JSON, but there may be other options

I'd therefore suggest that the problem of how to embed binary data in an event payload is not specific to meta operations, but they may add an extra layer of needed metadata.

> or SSE in core profile but I am not 100% sure

Note that SSE has a specific field for event name, so that doesn't cause a problem on its own, e.g. in the Core Profile I've defined this as being separate from the JSON payload (which is currently a direct mapping of the event data schema serialised in JSON).

In conclusion:
1. If you assume that each event is delivered in a separate payload then there's no need to provide a schema for delivering multiple events at once
2. When receiving events via a single endpoint (subscribed to using a `subscribeallevents` operation)  it _is_ necessary to distinguish between event types by their name. For mechanisms which don't have a native mechanism for this (like SSE) it may be necessary to define a schema to include the event name in a wrapper. I can't think of an obvious way to do that in the top level form of a meta operation, since as you say receiving an event is not an operation.
3. Generally speaking if you want to embed binary data in a JSON wrapper then you have to base64 encode it
4. For some event mechanisms it may be necessary to describe a different payload for a `subscribeallevents` request, but I haven't personally experienced this use case (for WebSockets and SSE all the required information is in the URL and headers)
